frontend/packages/network-attachment-definition-plugin/src/components/network-attachment-definitions/NetworkAttachmentDefinitionsForm.tsx

```diff
@@ -0,0 +1,433 @@
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+import * as classNames from 'classnames';
+import * as _ from 'lodash';
+import { Form, FormControl, FormGroup, HelpBlock } from 'patternfly-react';
+import { ActionGroup, Button } from '@patternfly/react-core';
+import {
+  ButtonBar,
+  Dropdown,
+  Firehose,
+  FirehoseResource,
+  history,
+} from '@console/internal/components/utils';
+import { k8sCreate } from '@console/internal/module/k8s';
+import * as k8sActions from '@console/internal/actions/k8s';
+import { validateDNS1123SubdomainValue, ValidationErrorType } from '@console/shared';
+import { NetworkAttachmentDefinitionModel, SriovNetworkNodePolicyModel } from '../..';
+import { NetworkAttachmentDefinitionConfig } from '../../types';
+import { ELEMENT_TYPES, NetworkTypeParams, networkTypeParams, networkTypes } from '../../constants';
+
+const NetworkAttachmentDefinitionFormBase = (props) => {
+  const { loaded, match, resources, hasSriovNetNodePolicyCRD } = props;
+  const namespace = _.get(match, 'params.ns', 'default');
+  const sriovNetNodePoliciesData = _.get(resources, 'sriovnetworknodepolicies.data', []);
+
+  const [loading, setLoading] = React.useState(loaded);
+  const [name, setName] = React.useState('');
+  const [description, setDescription] = React.useState('');
+  const [networkType, setNetworkType] = React.useState(null);
+  const [typeParamsData, setTypeParamsData] = React.useState({}); // TODO add typing
+  const [error, setError] = React.useState(null);
+  const [fieldErrors, setFieldErrors] = React.useState<FieldErrors>({});
+
+  const getNetworkTypeSpecificElements = React.useCallback(() => {
+    const handleTypeParamChange = (paramKey, event, elemType) => {
```

Comment:
all functions inside function are recreated everytime the parent function is called. Can you please move them outside ?

Reply:
Done!